File: faq-plus/configapp/Microsoft.Teams.Apps.FAQPlusPlus.Common/Providers/TicketsProvider.cs
// <copyright file="TicketsProvider.cs" company="Microsoft">
// Copyright (c) Microsoft. All rights reserved.
// </copyright>
namespace Microsoft.Teams.Apps.FAQPlusPlus.Common.Providers
{
    using System;
    using System.Threading.Tasks;
    using Microsoft.Teams.Apps.FAQPlusPlus.Common.Exceptions;
    using Microsoft.Teams.Apps.FAQPlusPlus.Common.Models;
    using Microsoft.WindowsAzure.Storage;
    using Microsoft.WindowsAzure.Storage.Table;

    /// <summary>
    /// Ticket provider helps in fetching and storing information in storage table.
    /// </summary>
    public class TicketsProvider : ITicketsProvider
    {
        private const string PartitionKey = "TicketInfo";
        private readonly Lazy<Task> initializeTask;
        private CloudTable ticketCloudTable;

        /// <summary>
        /// Initializes a new instance of the <see cref="TicketsProvider"/> class.
        /// </summary>
        /// <param name="connectionString">connection string of storage provided by dependency injection.</param>
        public TicketsProvider(string connectionString)
        {
            this.initializeTask = new Lazy<Task>(() => this.InitializeTableStorageAsync(connectionString));
        }

        /// <summary>
        /// Store or update ticket entity in table storage.
        /// </summary>
        /// <param name="ticket">Represents ticket entity used for storage and retrieval.</param>
        /// <returns><see cref="Task"/> that represents configuration entity is saved or updated.</returns>
        public Task UpsertTicketAsync(TicketEntity ticket)
        {
            ticket.PartitionKey = PartitionKey;
            ticket.RowKey = ticket.TicketId;

            if (ticket.Status > (int)TicketState.MaxValue)
            {
                throw new TicketValidationException($"The ticket status ({ticket.Status}) is not valid.");
            }

            return this.StoreOrUpdateTicketEntityAsync(ticket);
        }

        /// <summary>
        /// Get already saved entity detail from storage table.
        /// </summary>
        /// <param name="ticketId">ticket id received from bot based on which appropriate row data will be fetched.</param>
        /// <returns><see cref="Task"/> Already saved entity detail.</returns>
        public async Task<TicketEntity> GetTicketAsync(string ticketId)
        {
            await this.EnsureInitializedAsync().ConfigureAwait(false); // When there is no ticket created by end user and messaging extension is open by SME, table initialization is required before creating search index or datasource or indexer.
            if (string.IsNullOrEmpty(ticketId))
            {
                return null;
            }

            var searchOperation = TableOperation.Retrieve<TicketEntity>(PartitionKey, ticketId);
            var searchResult = await this.ticketCloudTable.ExecuteAsync(searchOperation).ConfigureAwait(false);

            return (TicketEntity)searchResult.Result;
        }

        /// <summary>
        /// Initialization of InitializeAsync method which will help in creating table.
        /// </summary>
        /// <returns>Represent a task with initialized connection data.</returns>
        private async Task EnsureInitializedAsync()
        {
            await this.initializeTask.Value.ConfigureAwait(false);
        }

        /// <summary>
        /// Create tickets table if it doesn't exist.
        /// </summary>
        /// <param name="connectionString">storage account connection string.</param>
        /// <returns><see cref="Task"/> representing the asynchronous operation task which represents table is created if its not existing.</returns>
        private async Task InitializeTableStorageAsync(string connectionString)
        {
            CloudStorageAccount storageAccount = CloudStorageAccount.Parse(connectionString);
            CloudTableClient cloudTableClient = storageAccount.CreateCloudTableClient();
            this.ticketCloudTable = cloudTableClient.GetTableReference(Constants.TicketTableName);

            await this.ticketCloudTable.CreateIfNotExistsAsync().ConfigureAwait(false);
        }

        /// <summary>
        /// Store or update ticket entity in table storage.
        /// </summary>
        /// <param name="entity">Represents ticket entity used for storage and retrieval.</param>
        /// <returns><see cref="Task"/> that represents configuration entity is saved or updated.</returns>
        private async Task<TableResult> StoreOrUpdateTicketEntityAsync(TicketEntity entity)
        {
            await this.EnsureInitializedAsync().ConfigureAwait(false);
            TableOperation addOrUpdateOperation = TableOperation.InsertOrReplace(entity);
            return await this.ticketCloudTable.ExecuteAsync(addOrUpdateOperation).ConfigureAwait(false);
        }
    }
}

File: faq-plus/bot/providers/ticketsProvider.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { TableService, createTableService } from 'azure-storage';
import { TicketEntity } from '../models/ticketEntity';
import { TicketState } from "../models/ticketState";

export class TicketsProvider {
    private readonly partitionKey: string = "TicketInfo";
    private readonly ticketTableName: string = "Tickets";

    private readonly connectionString: string;
    private tableService: TableService;

    constructor(connectionString: string) {
        this.connectionString = connectionString;
    }

    public async upsertTicket(ticket: TicketEntity): Promise<void> {
        ticket.PartitionKey = this.partitionKey;
        ticket.RowKey = ticket.TicketId;

        if (ticket.Status > TicketState.MaxValue) {
            throw new Error(`The ticket status (${ticket.Status}) is not valid.`);
        }

        await this.storeOrUpdateTicketEntity(ticket);
    }

    public async getTicket(ticketId: string): Promise<TicketEntity> {
        await this.ensureInitialize();
        if (!ticketId) {
            return null;
        }

        return new Promise((resolve, reject) => {
            this.tableService.retrieveEntity(this.ticketTableName, this.partitionKey,
                ticketId, (error, result: any) => {
                    if (error) {
                        reject(error);
                    }
                    let ticket = new TicketEntity();
                    const props = Object.keys(result);
                    props.forEach((key)=>{
                        ticket[key] = result[key]._;
                    })
                    resolve(ticket);
                });
        })
    }

    private async storeOrUpdateTicketEntity(entity: TicketEntity): Promise<void> {
        await this.ensureInitialize();

        return new Promise((resolve, rejcet) => {
            this.tableService.insertOrReplaceEntity(this.ticketTableName, entity,
                (error, result, response) => {
                    if (!error) {
                        resolve();
                    }
                    rejcet(error);
                });
        })
    }

    private async ensureInitialize(): Promise<void> {
        return new Promise((resolve, reject) => {
            if (this.tableService) {
                resolve();
            } else {
                this.tableService = createTableService(this.connectionString);
                this.tableService.createTableIfNotExists(this.ticketTableName, (error, result, response) => {
                    if (!error) {
                        resolve();
                    }
                    reject(error);
                })
            };
        })
    }
}